buildtools/npm/manifest.go

```diff
@@ -1,28 +1,64 @@
 package npm
 
 import (
+	"path/filepath"
+
 	"github.com/fossas/fossa-cli/errors"
 	"github.com/fossas/fossa-cli/files"
+	"github.com/fossas/fossa-cli/graph"
+	"github.com/fossas/fossa-cli/pkg"
 )
 
-type Manifest struct {
-	Name    string
-	Version string
-
+type manifest struct {
+	Name         string
+	Version      string
 	Dependencies map[string]string
 }
 
-func FromManifest(filename string) (Manifest, error) {
-	var manifest Manifest
-	err := files.ReadJSON(&manifest, filename)
+// PackageFromManifest generates a package definition for the provided manifest in the supplied directory
+func PackageFromManifest(pathElems ...string) (pkg.Package, error) {
```

Comment:
I think it makes more sense for `PackageFromManifest` to contain the direct import resolution logic. This has two advantages:

1. Consumers don't need to remember the unstated contract that `PackageFromManifest` only provides unresolved direct imports, and that they need to do resolution separately. A simpler interface is less error-prone.
2. This will make the logic in `FromNodeModules` more readable. The existing logic looks something like:

	```
	func nodeModulesHelper(manifest):
		pkg := PackageFromManifest(manifest)
	
		for import in unresolvedPkg.Imports:
			1. resolve import
			2. set resolved import field in pkg
			3. recurse nodeModulesHelper(import filepath)
	
		return deps
	```
	
	The new logic can look something like:
	```
	func nodeModulesHelper(manifest):
		// the imports in here will already be resolved
		pkg := PackageFromManifest(manifest)
	
		for import in pkg.Imports:
			recurse nodeModulesHelper(import filepath)
		
		return deps
	```

	You can store the import filepath in the `Location` field of the `pkg.ID` of the `pkg.Import.Resolved` during `PackageFromManifest` to avoid doing that WalkUp twice, although I think doing it twice (i.e. invoking `modulePath` once in `PackageFromManifest` and again in `nodeModulesHelper`) is acceptable performance overhead for the increased clarity.

If you disagree, I'd still like to add comments to the individual parts of logic in `nodeModulesHelper` (i.e. resolving, setting the field, recursing). Go lint doesn't _require_ comments on private functions, but we still _should_ comment them in cases like this where it clarifies the function's intention. Function names are not always sufficient for clarity, and sometimes it's better to have a shorter name and a clear comment for ergonomic reasons.

Reply:
I disagree here. `PackageFromManifest` is used in places where that resolution might not be possible (i.e. no node_modules folder; e.g. https://github.com/fossas/fossa-cli/pull/261/files#diff-612d68e420391fb3cc54bf19ef4b44f4R199).

I'll move forward with comments

Reply:
1. I don't see how resolution is not possible here. It's just not needed.
2. Maybe we should have different functions and data structures for "read a manifest" vs. "construct a package". I don't like how the constructed package implicitly has unresolved dependencies. Are there any cases where resolution is not needed _and_ we need a `pkg.Package`?

Reply:
The resolution not being possible would be in the case of not having `node_modules` folder. If it requires a . `node_modules` folder, then the function would error out inside the `IsBuilt` function.  

I think a fair middleground would be attempting to resolve with node_modules if possible

Reply:
You never _know_ if you have a `node_modules` folder. Even during the analysis, an unbuilt project is a failure mode we need to gracefully handle.